File: frontend/src/renderer/src/types/voice.ts
/** Voice-related types for STT, TTS, and audio streaming. */

// ---------------------------------------------------------------------------
// State
// ---------------------------------------------------------------------------

/** Full voice state tracked by the voice store. */
export interface VoiceState {
  /** Whether the microphone is currently recording. */
  isListening: boolean
  /** Whether STT is processing audio. */
  isProcessing: boolean
  /** Whether TTS is playing audio. */
  isSpeaking: boolean
  /** The latest transcript from STT. */
  transcript: string
  /** Microphone input level (0-1). */
  audioLevel: number
  /** Whether voice services are enabled in config. */
  enabled: boolean
  /** Microphone permission status. */
  micPermission: 'granted' | 'denied' | 'prompt'
  /** Voice WebSocket connection status. */
  connected: boolean
  /** Recording duration in seconds. */
  recordingDuration: number
}

// ---------------------------------------------------------------------------
// Audio
// ---------------------------------------------------------------------------

/** A chunk of audio data for streaming. */
export interface AudioChunk {
  /** Raw audio bytes. */
  data: ArrayBuffer
  /** Sample rate in Hz. */
  sampleRate: number
  /** Timestamp in milliseconds. */
  timestamp: number
}

// ---------------------------------------------------------------------------
// STT
// ---------------------------------------------------------------------------

/** Configuration for speech-to-text. */
export interface STTConfig {
  enabled: boolean
  engine: string
  model: string
  language: string
  device: string
  vad_filter: boolean
}

/** Result from STT transcription. */
export interface STTResult {
  /** Transcribed text. */
  text: string
  /** Detected language code. */
  language: string
  /** Confidence score (0.0–1.0, derived from model log-probabilities). */
  confidence: number
  /** Audio duration in seconds. */
  duration_s: number
  /** Whether this is a final result (vs. partial). */
  isFinal?: boolean
}

// ---------------------------------------------------------------------------
// TTS
// ---------------------------------------------------------------------------

/** Configuration for text-to-speech. */
export interface TTSConfig {
  enabled: boolean
  engine: 'piper' | 'xtts'
  voice: string
  speed: number
  sample_rate: number
}

// ---------------------------------------------------------------------------
// Voice WebSocket Messages
// ---------------------------------------------------------------------------

/** Client → Server: start recording. */
export interface VoiceStartMessage {
  type: 'voice_start'
  sample_rate?: number
}

/** Client → Server: stop recording. */
export interface VoiceStopMessage {
  type: 'voice_stop'
}

/** Client → Server: request TTS synthesis. */
export interface TTSSpeakMessage {
  type: 'tts_speak'
  text: string
}

/** Client → Server: cancel TTS playback. */
export interface TTSCancelMessage {
  type: 'tts_cancel'
}

/** Server → Client: voice ready status. */
export interface VoiceReadyMessage {
  type: 'voice_ready'
  stt_available: boolean
  tts_available: boolean
  stt_model?: string | null
  stt_engine?: string | null
  tts_engine?: string | null
  sample_rate?: number | null
}

/** Server → Client: recording started acknowledgment. */
export interface RecordingStartedMessage {
  type: 'recording_started'
}

/** Server → Client: recording stopped, possibly empty. */
export interface RecordingStoppedMessage {
  type: 'recording_stopped'
  empty: boolean
}

/** Server → Client: STT is processing. */
export interface STTProcessingMessage {
  type: 'stt_processing'
}

/** Server → Client: transcription result. */
export interface TranscriptMessage {
  type: 'transcript'
  text: string
  language: string
  confidence: number
  duration_s: number
}

/** Server → Client: TTS synthesis started. */
export interface TTSStartMessage {
  type: 'tts_start'
}

/** Server → Client: TTS synthesis complete. */
export interface TTSDoneMessage {
  type: 'tts_done'
}

/** Server → Client: TTS cancelled. */
export interface TTSCancelledMessage {
  type: 'tts_cancelled'
}

/** Server → Client: voice error. */
export interface VoiceErrorMessage {
  type: 'voice_error'
  message: string
}

/** Union of all voice WS messages from server. */
export type VoiceServerMessage =
  | VoiceReadyMessage
  | RecordingStartedMessage
  | RecordingStoppedMessage
  | STTProcessingMessage
  | TranscriptMessage
  | TTSStartMessage
  | TTSDoneMessage
  | TTSCancelledMessage
  | VoiceErrorMessage

/** Union of all voice WS messages from client. */
export type VoiceClientMessage =
  | VoiceStartMessage
  | VoiceStopMessage
  | TTSSpeakMessage
  | TTSCancelMessage

// ---------------------------------------------------------------------------
// Voice status (REST)
// ---------------------------------------------------------------------------

/** Response from GET /api/voice/status */
export interface VoiceStatusResponse {
  stt_available: boolean
  tts_available: boolean
  active_connections: number
}

File: backend/api/routes/voice.py
"""O.M.N.I.A. — Voice WebSocket endpoint (STT + TTS streaming)."""

from __future__ import annotations

import asyncio
import io
import json
import uuid
import wave
from collections import defaultdict

from fastapi import APIRouter, Request, WebSocket, WebSocketDisconnect
from loguru import logger
from starlette.websockets import WebSocketState

from backend.core.context import AppContext
from backend.core.event_bus import OmniaEvent
from backend.services.audio_utils import MAX_AUDIO_SIZE_BYTES

router = APIRouter(prefix="/voice", tags=["voice"])

# Active voice connections per IP (rate limiting).
_voice_connections: dict[str, int] = defaultdict(int)
_voice_lock = asyncio.Lock()
_MAX_VOICE_PER_IP = 2


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _ctx(ws_or_request: WebSocket | Request) -> AppContext:
    """Extract ``AppContext`` from ASGI app state."""
    return ws_or_request.app.state.context


async def _send_json(ws: WebSocket, data: dict) -> None:
    """Send JSON to client; silently ignore if connection already closed."""
    if ws.client_state == WebSocketState.CONNECTED:
        try:
            await ws.send_json(data)
        except Exception:
            pass


async def _stream_tts(
    ws: WebSocket, tts, text: str, cancel: asyncio.Event,
) -> None:
    """Stream TTS audio chunks (binary frames) to the client."""
    try:
        await _send_json(ws, {"type": "tts_start"})
        async for chunk in tts.synthesize_stream(text):
            if cancel.is_set():
                break
            if ws.client_state == WebSocketState.CONNECTED:
                await ws.send_bytes(chunk)
        await _send_json(ws, {"type": "tts_done"})
    except asyncio.CancelledError:
        await _send_json(ws, {"type": "tts_cancelled"})
    except Exception as exc:
        logger.error("TTS streaming error: {}", exc)
        await _send_json(ws, {"type": "voice_error", "message": "TTS synthesis error"})


async def _cancel_tts(
    tts_task: asyncio.Task | None, cancel_event: asyncio.Event,
) -> asyncio.Event:
    """Cancel an ongoing TTS task and return a fresh cancel event."""
    cancel_event.set()
    if tts_task and not tts_task.done():
        tts_task.cancel()
        try:
            await asyncio.wait_for(asyncio.shield(tts_task), timeout=1.0)
        except (asyncio.CancelledError, asyncio.TimeoutError, Exception):
            pass
    return asyncio.Event()


# ---------------------------------------------------------------------------
# WebSocket endpoint
# ---------------------------------------------------------------------------


@router.websocket("/ws/voice")
async def ws_voice(websocket: WebSocket) -> None:
    """Voice streaming WebSocket endpoint.

    Protocol (binary + JSON on the same connection):

    **Client → Server**
      - ``{"type": "voice_start"}`` — begin recording
      - binary frames — raw audio (PCM 16 kHz 16-bit mono)
      - ``{"type": "voice_stop"}`` — end recording, trigger STT
      - ``{"type": "tts_speak", "text": "..."}`` — request TTS synthesis
      - ``{"type": "tts_cancel"}`` — interrupt ongoing TTS playback
      - ``{"type": "ping"}``

    **Server → Client**
      - ``{"type": "voice_ready", "stt_available": bool, "tts_available": bool, "stt_model": str|null, "stt_engine": str|null, "tts_engine": str|null, "sample_rate": int|null}``
      - ``{"type": "recording_started"}``
      - ``{"type": "stt_processing"}``
      - ``{"type": "transcript", "text": "...", "language": "...", ...}``
      - ``{"type": "tts_start"}`` / ``{"type": "tts_done"}`` / binary TTS audio
      - ``{"type": "voice_error", "message": "..."}``
    """
    ctx = _ctx(websocket)
    client_ip = websocket.client.host if websocket.client else "unknown"
    session_id = uuid.uuid4().hex[:12]

    # Connection limit check.
    async with _voice_lock:
        if _voice_connections.get(client_ip, 0) >= _MAX_VOICE_PER_IP:
            await websocket.accept()
            await websocket.close(code=4029, reason="Too many voice connections")
            return
        await websocket.accept()
        _voice_connections[client_ip] += 1
    logger.debug("Voice WS connected: session={} ip={}", session_id, client_ip)
    await ctx.event_bus.emit(OmniaEvent.VOICE_SESSION_START, session_id=session_id)

    # Per-session state.
    audio_buffer = bytearray()
    is_recording = False
    client_sample_rate = 16000  # default; updated by voice_start
    cancel_event = asyncio.Event()
    tts_task: asyncio.Task | None = None

    try:
        stt = ctx.stt_service
        tts = ctx.tts_service

        # Inform client about service availability (graceful degradation).
        if not stt:
            await _send_json(websocket, {
                "type": "voice_error", "message": "STT service not available",
            })
        if not tts:
            await _send_json(websocket, {
                "type": "voice_error", "message": "TTS service not available",
            })
        await _send_json(websocket, {
            "type": "voice_ready",
            "stt_available": stt is not None,
            "tts_available": tts is not None,
            "stt_model": stt.model_name if stt else None,
            "stt_engine": stt.engine if stt else None,
            "tts_engine": tts.engine if tts else None,
            "sample_rate": tts.sample_rate if tts else None,
        })

        while True:
            message = await websocket.receive()

            if message.get("type") == "websocket.disconnect":
                break

            # --- Binary frame: audio data ----------------------------------
            if "bytes" in message:
                if is_recording and stt:
                    if len(audio_buffer) + len(message["bytes"]) > MAX_AUDIO_SIZE_BYTES:
                        is_recording = False
                        audio_buffer.clear()
                        await _send_json(websocket, {
                            "type": "voice_error",
                            "message": "Audio buffer limit exceeded",
                        })
                    else:
                        audio_buffer.extend(message["bytes"])
                continue

            # --- JSON frame ------------------------------------------------
            if "text" not in message:
                continue
            try:
                data = json.loads(message["text"])
            except json.JSONDecodeError:
                continue

            msg_type = data.get("type", "")

            if msg_type == "voice_start":
                # Auto-cancel ongoing TTS when user starts speaking.
                cancel_event = await _cancel_tts(tts_task, cancel_event)
                audio_buffer.clear()
                is_recording = True
                # Client may send the actual AudioContext sample rate.
                try:
                    client_sample_rate = max(8000, min(48000, int(data.get("sample_rate", 16000))))
                except (ValueError, TypeError):
                    client_sample_rate = 16000
                logger.debug(
                    "Recording started: session={} sample_rate={}",
                    session_id, client_sample_rate,
                )
                await _send_json(websocket, {"type": "recording_started"})

            elif msg_type == "voice_stop":
                is_recording = False
                if not stt or len(audio_buffer) == 0:
                    await _send_json(websocket, {
                        "type": "recording_stopped", "empty": True,
                    })
                    continue

                audio_copy = bytes(audio_buffer)
                audio_buffer.clear()
                await _send_json(websocket, {"type": "recording_stopped", "empty": False})
                await _send_json(websocket, {"type": "stt_processing"})

                # Wrap raw PCM-16 mono in a WAV container using the
                # actual sample rate reported by the client.
                logger.debug(
                    "Wrapping PCM buffer: {} bytes, sample_rate={}",
                    len(audio_copy), client_sample_rate,
                )
                wav_buf = io.BytesIO()
                with wave.open(wav_buf, "wb") as wf:
                    wf.setnchannels(1)
                    wf.setsampwidth(2)
                    wf.setframerate(client_sample_rate)
                    wf.writeframes(audio_copy)
                wav_bytes = wav_buf.getvalue()

                try:
                    await ctx.event_bus.emit(OmniaEvent.STT_STARTED, session_id=session_id)
                    result = await stt.transcribe(wav_bytes)
                    logger.debug(
                        "STT result: text={!r} lang={} conf={:.3f} dur={:.1f}s",
                        result.text, result.language,
                        result.confidence, result.duration_s,
                    )
                    await ctx.event_bus.emit(
                        OmniaEvent.STT_RESULT,
                        session_id=session_id, text=result.text,
                    )
                    await _send_json(websocket, {
                        "type": "transcript",
                        "text": result.text,
                        "language": result.language,
                        "confidence": result.confidence,
                        "duration_s": result.duration_s,
                    })
                except RuntimeError as exc:
                    logger.error(
                        "STT RuntimeError (model will reload on next call): {}",
                        exc,
                    )
                    await ctx.event_bus.emit(
                        OmniaEvent.STT_ERROR,
                        session_id=session_id, error=str(exc),
                    )
                    await _send_json(websocket, {
                        "type": "voice_error",
                        "message": f"Transcription failed: {type(exc).__name__}. "
                                   "The model will reload on the next attempt.",
                    })
                except Exception as exc:
                    logger.exception("STT transcription failed: {}", exc)
                    await ctx.event_bus.emit(
                        OmniaEvent.STT_ERROR,
                        session_id=session_id, error=str(exc),
                    )
                    await _send_json(websocket, {
                        "type": "voice_error",
                        "message": f"Transcription failed: {type(exc).__name__}",
                    })

            elif msg_type == "tts_speak":
                text = data.get("text", "").strip()
                if not tts or not text:
                    continue
                if len(text) > 10_000:
                    await _send_json(websocket, {
                        "type": "voice_error",
                        "message": "Text too long for TTS synthesis",
                    })
                    continue
                cancel_event = await _cancel_tts(tts_task, cancel_event)
                tts_task = asyncio.create_task(
                    _stream_tts(websocket, tts, text, cancel_event),
                )

            elif msg_type == "tts_cancel":
                cancel_event = await _cancel_tts(tts_task, cancel_event)

            elif msg_type == "ping":
                await _send_json(websocket, {"type": "pong"})

    except WebSocketDisconnect:
        logger.debug("Voice WS disconnected: session={}", session_id)
    except Exception as exc:
        logger.error("Voice WS error [{}]: {}", session_id, exc)
    finally:
        async with _voice_lock:
            _voice_connections[client_ip] = max(
                0, _voice_connections.get(client_ip, 1) - 1,
            )
            if _voice_connections.get(client_ip, 0) <= 0:
                _voice_connections.pop(client_ip, None)
        cancel_event.set()
        if tts_task and not tts_task.done():
            tts_task.cancel()
        try:
            await ctx.event_bus.emit(
                OmniaEvent.VOICE_SESSION_END, session_id=session_id,
            )
        except Exception:
            pass


# ---------------------------------------------------------------------------
# REST — voice service health
# ---------------------------------------------------------------------------


@router.get("/status")
async def voice_status(request: Request) -> dict:
    """Return availability of STT/TTS services and active connection count."""
    ctx = _ctx(request)
    stt_ok = (
        ctx.stt_service is not None and await ctx.stt_service.health_check()
    )
    tts_ok = (
        ctx.tts_service is not None and await ctx.tts_service.health_check()
    )
    return {
        "stt_available": stt_ok,
        "tts_available": tts_ok,
        "active_connections": sum(_voice_connections.values()),
    }
